torch/jit/__init__.py

```diff
@@ -369,14 +369,18 @@ def _script_graph(fn, _frames_up=0):
 
 def script(fn, _frames_up=0):
     graph = _script_graph(fn, _frames_up=_frames_up + 1)
-    return torch._C.GraphExecutor(graph, True)
+    fn_attributes = {i: getattr(fn, i) for i in functools.WRAPPER_ASSIGNMENTS}
+    Wrapper = type('GraphExecutorWrapper', (torch._C.GraphExecutor, ), fn_attributes)
+    return Wrapper(graph, True)
 
 
 ScriptMethodStub = namedtuple('ScriptMethodStub', ('resolution_callback', 'ast'))
 
 
 def script_method(fn):
-    return ScriptMethodStub(createResolutionCallback(frames_up=1), get_jit_ast(fn))
+    fn_attributes = {i: getattr(fn, i) for i in functools.WRAPPER_ASSIGNMENTS}
+    Wrapper = type('ScriptMethodStubWrapper', (ScriptMethodStub, ), fn_attributes)
+    return Wrapper(createResolutionCallback(frames_up=1), get_jit_ast(fn))
```

Comment:
Actually, do we really need to go through this dynamic subclass creation? Wouldn't it be sufficient to just call `setattr` on the `ScriptMethodStub` instance, for all attrs from `WRAPPER_ASSIGNMENTS`? Also, can you please paste the exact error you're getting when you try to use `functools.wraps`?

Reply:
@apaszke Actually in my first commit of this PR https://github.com/pytorch/pytorch/pull/7898/commits/313a65c8c54a74c23ef6faa8b6ce7bd5f5de7489, I was using `functools.wraps`, but all the python 2.7 build fails. You can check at, for example:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py2.7.9-build/7788/console
The error message was:
```
15:43:49 Traceback (most recent call last):
15:43:49   File "build.py", line 2, in <module>
15:43:49     import torch
15:43:49   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/__init__.py", line 282, in <module>
15:43:49     import torch.jit
15:43:49   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/jit/__init__.py", line 721, in <module>
15:43:49     class _ConstSequential(_ConstModuleList):
15:43:49   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/jit/__init__.py", line 727, in _ConstSequential
15:43:49     @script_method
15:43:49   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/jit/__init__.py", line 384, in script_method
15:43:49     class Wrapper(ScriptMethodStub):
15:43:49   File "/opt/python/2.7.9/lib/python2.7/functools.py", line 33, in update_wrapper
15:43:49     setattr(wrapper, attr, getattr(wrapped, attr))
15:43:49 AttributeError: attribute '__doc__' of 'type' objects is not writable
15:43:49 + cleanup
```

Also, the following experiment would be helpful to demonstrate the problem:
```python
from __future__ import print_function
import functools
import platform
from collections import *

print(platform.python_version())
print(platform.python_implementation())
print(platform.python_compiler())

A = namedtuple('A', ('X', 'Y'))

def f():
    """a function"""
    pass

try:
    @functools.wraps(f, updated=())
    class B(A):
        pass

    print(B.__doc__)
except AttributeError as e:
    print('Error @functools.wraps:', e)

try:
    a = A(1,2)
    a.__doc__ = 'test'
except AttributeError as e:
    print('Error @setattr', e)
```
On my python2, I get:
```
2.7.10
CPython
GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)
Error @functools.wraps: attribute '__doc__' of 'type' objects is not writable
Error @setattr 'A' object attribute '__doc__' is read-only
```
And on python3, I get:
```
3.6.5
CPython
GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)
a function
Error @setattr 'A' object attribute '__doc__' is read-only
```